### PR TITLE
Import emscripten's relooper, port it to the binaryen AST, and provide a C API

### PR DESCRIPTION
**Pull request by kripken**
This turned out to be easier and nicer than I thought. Switching from JS output in text to wasm in AST was a nice change in itself. In addition, we don't need a bunch of relooper optimizations, since binaryen has AST control flow optimizations already, and we can just rely on those. So this is simpler and cleaner than the previous relooper.

Current limitations: no switches, and while basic tests pass, it should be tested a lot more.

### CODE REVIEW

**Comment by kripken** [comment]
@jfbastien: this is failing on ubsan, on

````
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/bits/stl_tree.h:741:25: runtime error: downcast of address 0x7fffe6af0888 with insufficient space for an object of type '_Rb_tree_node<std::pair<CFG::Block *const, CFG::Block *> >'
0x7fffe6af0888: note: pointer points here
 ff 7f 00 00  00 00 00 00 00 00 00 00  f0 a4 12 02 00 00 00 00  f0 a4 12 02 00 00 00 00  30 a6 12 02
              ^ 
````

(e.g. see [this log](https://api.travis-ci.org/jobs/127976646/log.txt?deansi=true)). No stack trace even in a debug build. And that class isn't something I defined, I guess it's part of the STL - so this must be hitting UB inside the STL somewhere? Any tips on how one would debug something like this?

Guessing that that's a map of `Block*` to `Block*`, I looked at all the code using that. There isn't any iteration on them, so no iterator invalidation I can see. Some `find()`s are done but they look safe too. Weird.

**Comment by binji** [comment]
@kripken A quick googling found this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60734

**Comment by kripken** [comment]
Oh wow, thanks @binji! I didn't even think to look for that...

Ok, let's see if ignoring that specific file in ubsan helps.

**Comment by kripken** [comment]
Ok great, adding that stdc++ file to the ubsan list avoids the issue, so this should be good to go.

File: ubsan.blacklist [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # Work around libstdc++ bug: https://llvm.org/bugs/show_bug.cgi?id=18156
 # Also see: http://lists.llvm.org/pipermail/cfe-dev/2015-January/040945.html
 src:*/ios_base.h
+# Work around another libstdc++ bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60734
+src:*/stl_tree.h
```
Comment (jfbastien) [inline ubsan.blacklist]: Sorry, only just catching up to github. This looks good.

**Pull request merged**